lib/charms/storage_libs/v0/cephfs_interfaces.py

```diff
@@ -0,0 +1,358 @@
+    CharmBase,
+    CharmEvents,
+    RelationChangedEvent,
+    RelationDepartedEvent,
+    RelationEvent,
+    RelationJoinedEvent,
+)
+from ops.framework import EventSource, Object
+from ops.model import Relation
+
+# The unique Charmhub library identifier, never change it
+LIBID = "UNPUBLISHED"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+_logger = logging.getLogger(__name__)
+
+@dataclass(frozen=True)
+class _Transaction:
+    """Store transaction information between to data mappings."""
+
+    added: Set
+    changed: Set
+    deleted: Set
+
+
+def _eval(event: RelationChangedEvent, bucket: str) -> _Transaction:
+    """Evaluate the difference between data in an integration changed databag.
+
+    Args:
+        event: Integration changed event.
+        bucket: Bucket of the databag. Can be application or unit databag.
+
+    Returns:
+        _Transaction:
+            Transaction info containing the added, deleted, and changed
+            keys from the event integration databag.
+    """
+    # Retrieve the old data from the data key in the application integration databag.
+    old_data = json.loads(event.relation.data[bucket].get("cache", "{}"))
+    # Retrieve the new data from the event integration databag.
+    new_data = {
+        key: value for key, value in event.relation.data[event.app].items() if key != "cache"
+    }
+    # These are the keys that were added to the databag and triggered this event.
+    added = new_data.keys() - old_data.keys()
+    # These are the keys that were removed from the databag and triggered this event.
+    deleted = old_data.keys() - new_data.keys()
+    # These are the keys that already existed in the databag, but had their values changed.
+    changed = {key for key in old_data.keys() & new_data.keys() if old_data[key] != new_data[key]}
+    # Convert the new_data to a serializable format and save it for a next diff check.
+    event.relation.data[bucket].update({"cache": json.dumps(new_data)})
+
+    # Return the transaction with all possible changes.
+    return _Transaction(added, changed, deleted)
+
+
+class ServerConnectedEvent(RelationEvent):
+    """Emit when a CephFS server is integrated with CephFS client."""
+
+@dataclass
+class CephFSAuthInfo:
+    """Authorization info to access a CephFS share."""
+    username: str
+    """Name of the user authorized to access the CephFS share."""
+    key: str
+    """Cephx key for the provided user."""
+
+    dict = asdict
+
+@dataclass(init=False)
+class CephFSShareInfo:
+    """Information about a shared CephFS."""
+    fsid: str
+    """Id of the Ceph cluster."""
+    name: str
+    """Name of the exported CephFS."""
+    path: str
+    """Exported path of the CephFS."""
+    monitor_hosts: [str]
+    """Address list of the available Ceph MON nodes."""
```

Comment:
Similar to my earlier comment. We can just roll the docstring for each attribute into the class docstring:

```suggestion
    """Information about a shared CephFS.
    
    Attributes:
        fsid: ID of the Ceph cluster.
        name: Name of the exported CephFS share.
        path: Exported path of the CephFS share.
        monitor_hosts: Address list of the available Ceph MON nodes.
    """
    fsid: str
    name: str
    path: str
    monitor_hosts: [str]
```